### PR TITLE
[Coverage] Rework !SystemHeadersCoverage

### LINKED ISSUE #78920: Weird Clang's branch coverage with macros from system headers
Clang eliminates whole MACRO expr in branch coverage.

(https://godbolt.org/z/3o9MTf4eW)
```
/* header.h */
#define ISN(c) ('0' <= (c) && (c) <= '9')
```

```
#include <header.h>
int isx(int c) {
    return (ISN(c) || ('A' <= c && c <= 'F'));
}
```
```
clang -fprofile-instr-generate -fcoverage-mapping -Xclang -dump-coverage-mapping -isystem .
```
The emission of the record is as below; `ISN(c)` cannot be seen nor evaluated.
Could I see the condition of `ISN(c)`?
```
isx:
  File 0, 9:16 -> 11:2 = #0
  File 0, 10:23 -> 10:45 = #1
  File 0, 10:24 -> 10:32 = #1
  Branch,File 0, 10:24 -> 10:32 = #4, (#1 - #4)
  File 0, 10:36 -> 10:44 = #4
  Branch,File 0, 10:36 -> 10:44 = #5, (#4 - #5)
```

WIth `-fcoverage-mcdc`, `llvm-cov` crashes. (cc: @evodius96)
A workaround is adding `-mllvm -system-headers-coverage` to `clang`.

### PR DESCRIPTION
**Pull request by chapuni**
- Introduce `LeafExprSet`,
  - Suppress traversing LAnd and LOr expr under system headers.
  - Handle LAnd and LOr as instrumented leaves to override `!isInstrumentedCondition(C)`.
- Replace Loc with FileLoc if it is expanded with system headers.

Fixes #78920

### CODE REVIEW

File: clang/lib/CodeGen/CoverageMappingGen.cpp [hunk 1]
```diff
@@ -2064,7 +2082,20 @@ struct CounterCoverageMappingBuilder
     createDecisionRegion(E, DecisionParams);
   }
 
+  /// Check if E belongs to system headers.
+  bool isExprInSystemHeader(const BinaryOperator *E) const {
```
Comment (ornata) [inline clang/lib/CodeGen/CoverageMappingGen.cpp:2086]: assert E is not nullptr?

Comment (chapuni) [reply clang/lib/CodeGen/CoverageMappingGen.cpp:2086]: I assume each visitor (`visitBinLAnd` `visitBinLOr`) will not receive nullptr.

**Review by ornata — APPROVED** [review]
LGTM

**Review by hanickadot — APPROVED** [review]
LGTM

**Comment by AaronBallman** [comment]
FWIW, I landed 3591da9f1ccbd8b19fef4814f96638dbbe9c2b40 to address testing issues on Windows.

**Comment by chapuni** [comment]
@AaronBallman Thanks. Aha, I've noticed my test was incompatible for targeting msvc mangling.

**Pull request merged**